_posts/2016-10-18-using-rspecs-documentation-format.markdown

```diff
@@ -0,0 +1,63 @@
+---
+layout: post
+title:  "Brief look at RSpec's formatting options"
+date: 2016-10-18 14:29:00
+categories: ["rspec", "ruby"]
+author: "mauro-oto"
+---
+
+A few weeks ago, I noticed weird output in the RSpec test suite (~4000 tests)
+for a Rails application:
+
+```
+................................................................................
+................................................................................
+................................................................................
+.............unknown OID 353414: failed to recognize type of '<field>'. It will
+be treated as String............................................................
+................................................................................
+```
+
+This Rails app uses a PostgreSQL database. After some Googling, it turns out
+that this is a warning from PostgreSQL. When the database doesn't recognize the
+type to use for a column, it casts to string by default.
+
+<!--more-->
+
+I first thought I could look through previous CI runs to find out when this
+warning first started showing up. Then I decided to use RSpec's documentation
+format instead. I ran the build locally, and the test that triggers the warning
+for the first time was exposed:
+
+```
+➜  myproject git:(develop) ✗ bundle exec rspec spec --format documentation
+
+...
+
+Admin::ClientsController
+  GET 'index'
+unknown OID 353414: failed to recognize type of 'client_campaigns'. It will be treated as String.
+    non-admin user
+      does not authorize non-admin users
+    admin user
+      lists all clients
+
+...
+
+```
+
+The RSpec documentation format shows us the test description and the example
+names. Because of this, we can tell what test we can run to reproduce the
+warning. It also allows us to figure out where in the codebase the warning is
+triggered from.
+
+A good alternative to the documentation format is the [Fuubar gem](https://github.com/thekompanee/fuubar).
+Using `rspec --format Fuubar`, you get a nice looking progress bar, as well as
+its coolest feature, insta-failing tests. With Fuubar, you don't have to wait
+until the build ends or until you ctrl+c out of RSpec, you get the failure
+result right away.
+
+Since `--format Fuubar` doesn't show test descriptions by default, you can
+combine Fuubar with the documentation format, using
+`rspec --format Fuubar --format documentation`. This way, you get both the
+documentation format and insta-failing tests.
```

Comment:
@etagwerker I'll add a comment about that, I didn't know it was a native option. However, there's a slight difference. Fail fast means the build stops after 1 or n tests fail, insta-failing means Fuubar shows the failed tests, the test failure output (what expectation was not met), and won't stop the build for you.